stripe-ui-core/src/main/java/com/stripe/android/uicore/image/StripeImageLoader.kt

```diff
@@ -0,0 +1,164 @@
+package com.stripe.android.uicore.image
+
+import android.content.Context
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import androidx.annotation.RestrictTo
+import androidx.annotation.WorkerThread
+import com.stripe.android.core.Logger
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+import java.net.URL
+import java.util.concurrent.ConcurrentHashMap
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+
+/**
+ * Image loader that fetches images from memory, disk or network and runs
+ * cache policy accordingly.
+ *
+ * [StripeImageLoader] is stateful as it holds the memoryCache instance. For
+ * memory cache to work the image loader instance needs to be shared.
+ *
+ * @param memoryCache, memory cache to be used, or null if no memory cache is desired.
+ * @param diskCache, memory cache to be used, or null if no memory cache is desired.
+ */
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+class StripeImageLoader(
+    context: Context,
+    private val logger: Logger,
+    private val memoryCache: ImageLruMemoryCache? = ImageLruMemoryCache(),
+    private val diskCache: ImageLruDiskCache? = ImageLruDiskCache(
+        context = context,
+        uniqueName = "stripe_image_cache"
+    ),
+) {
+
+    private val imageLoadMutexes = ConcurrentHashMap<String, Mutex>()
+
+    /**
+     * loads the given [url] with the associated [width]x[height].
+     *
+     * If the same [url] is being loaded concurrently, function will be suspended until
+     * the original load completes.
+     */
+    suspend fun load(
+        url: String,
+        width: Int,
+        height: Int
+    ): Result<Bitmap> = withContext(Dispatchers.IO) {
+        withMutexByUrlLock(url) {
+            loadFromMemory(url) ?: loadFromDisk(url) ?: loadFromNetwork(url, width, height)
+        }
+    }
+
+    private fun loadFromMemory(url: String): Result<Bitmap>? {
+        return memoryCache?.getBitmap(url)
+            .also {
+                if (it != null) {
+                    debug("Image loaded from memory cache")
+                } else {
+                    debug("Image not found on memory cache")
+                }
+            }
+            ?.let {
+                diskCache?.put(url, it)
+                Result.success(it)
+            }
+    }
+
+    private fun loadFromDisk(url: String): Result<Bitmap>? {
+        return diskCache?.getBitmap(url)
+            .also {
```

Comment:
We don't have a code style rule around this, but tend to simplify these functions that just return something as
```suggestion
    private fun loadFromDisk(url: String) = diskCache?.getBitmap(url).also {
    
    
```
With the type if it's not clear, but saving on the brackets and return.

Reply:
https://github.com/stripe/stripe-android/pull/5307/commits/59e6d0eb7522608395554fbfedbf3635cd83c209